### ISSUE TITLE
parallel schedule

### ISSUE THREAD

**Issue by Melisius** [numba/numba#2649]
The numba parellel with prange is awesome, but I often encounter for loops where the work needed to be done for each "i" in the loop is not the same. A scheduling option like the one in OpenMP would be a nice addition to the numba parallel feature.

For OpenMP this is descirbed as:

"dynamic: 
Use the internal work queue to give a chunk-sized block of loop iterations to each thread. When a thread is finished, it retrieves the next block of loop iterations from the top of the work queue. By default, the chunk size is 1. Be careful when using this scheduling type because of the extra overhead involved."

**Comment by seibert** [comment]
Yes, this is something on our radar.  Now that Numba has more multithreading capabilities, we are seeing the inadequacies of our simple thread pool.  We are hoping in early 2018 to improve our work scheduling system for this scenario and also to allow us to make better use of TBB.

**Comment by NicolasHug** [comment]
+1 for more scheduling options.

I think the one I need is the 'static' scheduling instead of 'dynamic' (I'm not familiar with openMP so I'm not sure)

Let's say I need to update an array with independent updates, in a parallel fashion, e.g.:

```python
array += 1  # broadcasting, not parallel
```

Instead of running:

```python
for i in prange(array.shape[0]):
     array[i] += 1
```

which adds a lot of overhead, I often find myself dividing the array into `n_threads` equally-sized chunks:

```python

@njit
def _get_threads_chunks(total_size):
    # Divide [0, total_size - 1] into n_threads contiguous regions, and
    # returns the starts and ends of each region. Used to simulate a 'static'
    # scheduling.
    n_threads = numba.config.NUMBA_DEFAULT_NUM_THREADS
    sizes = np.full(n_threads, total_size // n_threads, dtype=np.int32)
    sizes[:total_size % n_threads] += 1
    starts = np.zeros(n_threads, dtype=np.int32)
    starts[1:] = np.cumsum(sizes[:-1])
    ends = starts + sizes

    return starts, ends, n_threads

@njit(parallel=True)
def update(array):
    starts, ends, n_threads = _get_threads_chunks(total_size=array.shape[0])
    for thread_idx in prange(n_threads):
        for i in range(starts[thread_idx], ends[thread_idx]):
            array[i] += 1

```

It would be great to have this as a builtin feature! Thanks for a all the work on numba!

**Comment by ogrisel** [comment]
May I suggest ["guided" ](http://jakascorner.com/blog/2016/06/omp-for-scheduling.html#guided)as the default strategy? It should have the same performance benefits as "static" when data is contiguous but also mitigate stragglers.

But being able to pass scheduling hints (especially minimum chunk size for guided) would be very nice.

**Comment by ogrisel** [comment]
Actually I am not sure what should be the default :) We probably need some benchmarks.

**Comment by mikecroucher** [comment]
I found myself here after some benchmarking demonstrated that I am experiencing the exact problem described by @erikkjellgren.  

I have an explicit parallel loop using prange and the way that each iteration is distributed among cores by Numba is such that the code eventually ends up with one core doing all the work while all of the rest sit and watch.  This ruins scaling in what is otherwise an almost embarrassingly parallel problem for me.   

I should be able to fix this to a degree by chopping up my code a little more but having more scheduler options might solve problems such as this quite easily.  

As such, I am bumping this thread to ask 'What has been done here?'.  I can't find anything in the docs.

**Comment by DrTodd13** [comment]
@mikecroucher How many cores (C) do you have and how many iterations (I) do you have?  There was a recent bug fix for when I is only slightly larger than C and doesn't divide evenly.  I'm not even sure it is merged yet.  Is it your case that the computation per iteration varies wildly?  We could probably easily add an option to break up the parallel regions into M*N workgroups instead of N workgroups, where N is NUMBA_NUM_THREADS and M is a small constant like 5.  I think Numba's gufunc mechanism would then effectively do something like dynamic work-stealing of those work-groups and might balance the load a bit better.  @sklam @stuartarchibald

**Comment by mikecroucher** [comment]
@DrTodd13 Number of cores varies a lot since I use cloud instances.  Recent core counts I've used are 4 (laptop). 8 (cheap Azure instance), 44 (Azure's Intel-based HC instances) and 120 (Azure's AMD based HBv2 instances). Loop size for the application I have in mind is anywhere from 200 to 5000 and length of each iteration varies a lot. 

Numba has been instrumental in getting the performance needed and fixing this would help squeeze out the last few percent.

**Comment by DrTodd13** [comment]
> +1 for more scheduling options.
> 
> I think the one I need is the 'static' scheduling instead of 'dynamic' (I'm not familiar with openMP so I'm not sure)
> 
> Let's say I need to update an array with independent updates, in a parallel fashion, e.g.:
> 
> ```python
> array += 1  # broadcasting, not parallel
> ```
> 
> Instead of running:
> 
> ```python
> for i in prange(array.shape[0]):
>      array[i] += 1
> ```
> 
> which adds a lot of overhead, I often find myself dividing the array into `n_threads` equally-sized chunks:
> 
> ```python
> @njit
> def _get_threads_chunks(total_size):
>     # Divide [0, total_size - 1] into n_threads contiguous regions, and
>     # returns the starts and ends of each region. Used to simulate a 'static'
>     # scheduling.
>     n_threads = numba.config.NUMBA_DEFAULT_NUM_THREADS
>     sizes = np.full(n_threads, total_size // n_threads, dtype=np.int32)
>     sizes[:total_size % n_threads] += 1
>     starts = np.zeros(n_threads, dtype=np.int32)
>     starts[1:] = np.cumsum(sizes[:-1])
>     ends = starts + sizes
> 
>     return starts, ends, n_threads
> 
> @njit(parallel=True)
> def update(array):
>     starts, ends, n_threads = _get_threads_chunks(total_size=array.shape[0])
>     for thread_idx in prange(n_threads):
>         for i in range(starts[thread_idx], ends[thread_idx]):
>             array[i] += 1
> ```
> 
> It would be great to have this as a builtin feature! Thanks for a all the work on numba!

I don't understand this at all because this is already what we do internally.  This is the default Numba/ParallelAccerlator behavior.  There's isn't a lot of overhead by default as we don't do one iteration at a time.

**Comment by DrTodd13** [comment]
> @DrTodd13 Number of cores varies a lot since I use cloud instances. Recent core counts I've used are 4 (laptop). 8 (cheap Azure instance), 44 (Azure's Intel-based HC instances) and 120 (Azure's AMD based HBv2 instances). Loop size for the application I have in mind is anywhere from 200 to 5000 and length of each iteration varies a lot.
> 
> Numba has been instrumental in getting the performance needed and fixing this would help squeeze out the last few percent.

@mikecroucher If you are eager, you can try out the new PR I made associated with this issue and see if it helps you.  Let me know if you have any questions.
